### PR TITLE
Water ballast synch between XCV and XCS in absolute ballast mode

### LINKED ISSUE #824: XCVario to XCSoar ballast synchronisation errors
Hardware 
XCvario firmware v21.1230-17
Huawei P30 lite Android v10.0.0.491 XCSoar v7.21
Bi-directional data exchange over BT using XCV drivers and protocol
Earlier versions of XCS exhibit problem.

Problem
This may relate to issue #807
Loaded into XCV, dry mass, pilot weight, and polar reference mass expressed as wing loading kg/m2
and in XCS Planes, Wing Area, Ref Mass and Dry mass (which is to include pilot according to manual).
With zero litres of ballast on both XCV and XCS when I open the flight setup page on XCS there is 42L that has been loaded,  Closing the flight setup page writes back 42l to XCV, opening the flight page again 81l is then loaded to XCS which is then written back to XCV and so this continues each time I open the flight setup page,116l then 148l and so on. Pressing the Dump button in XCS produces some very interesting results!!!
Changing ballast in XCS to zero or any other number, writes back that number to XCV but the addition process starts again when XCS Flight Setup is opened.
It is not possible to zero or synchronise the ballast between XCS and XCV when bi-directional data exchange is running. The only solution is disable the bi directional data exchange "Sync from device" with the loss of the QNH,MC and Bugs sync, 
Adjusting the dry mass and the pilot weight to equal the ref mass in XCV also overcomes the problem but at the expense of STF on the XCV.

My Observation for what its worth!
Looking at the Cambridge protocol which I think is used, in part, in the Xcvario protocol, calls for a ballast ratio made up of the total mass of the glider (dry mass, pilot, water) divided by the polar ref mass). In the case of my setup this is 1.27 with no ballast (water).
Monitoring the data stream from XCV I see the ballast variable is indeed  1.27. 
XCS seems to be treating any ratio over 1 as additional ballast. 
If I set the total mass equal to the ref mass in XCV, as suggested above, I get a ballast ratio of one on the data stream and the problem goes away, STF on XCV not withstanding.

### PR DESCRIPTION
**Pull request by iltis42**
**Brief summary of the changes**
----------------------------
Migrate fractional ballast exchange to absolute values in XCVario protocol

- Implement new fields for absolute values crew weight and empty mass for both directions and for ballast from device
- Migrate .xcp file extension appending into PlaneGlue to support resue of FileWrite() and avoid duplicate code
- Add backward compatible XCVario protocol version handling, now having a protocol version 2 running in absolute ballast mode
- Save plane file when as well when synchronized by external settings

**Related issues and discussions**
------------------------------
Closes #824 
#1017
https://github.com/iltis42/XCVario/issues/202

### CODE REVIEW

File: src/Dialogs/Plane/PlaneListDialog.cpp
Comment (MaxKellermann) [inline src/Dialogs/Plane/PlaneListDialog.cpp:237]: Simpler:
```
  const auto filename = PlaneGlue::GetFileName(plane);
```
Using `auto` also allows changing the function's return type (e.g. larger buffer) without changing all the callers.

File: src/Dialogs/Plane/PlaneListDialog.cpp
Comment (MaxKellermann) [inline src/Dialogs/Plane/PlaneListDialog.cpp:280]: `auto`

File: src/Plane/PlaneFileGlue.cpp
Comment (MaxKellermann) [inline src/Plane/PlaneFileGlue.cpp:215]: Missing `noexcept`

File: src/Plane/PlaneFileGlue.cpp
Comment (MaxKellermann) [inline src/Plane/PlaneFileGlue.cpp:219]: Superfluous parantheses

File: src/Plane/PlaneFileGlue.cpp
Comment (MaxKellermann) [inline src/Plane/PlaneFileGlue.cpp:222]: API documentation should be in doxygen-style comments in the header

File: src/Plane/PlaneFileGlue.cpp
Comment (MaxKellermann) [inline src/Plane/PlaneFileGlue.cpp:227]: Simpler and more efficient:
```
  const auto path = LocalPath(GetFileName(plane));
```
This allows the compiler to generate code to free the stack space occupied by `GetFileName()` immediately after returning from `LocalPath()`.

File: src/Plane/PlaneFileGlue.cpp
Comment (MaxKellermann) [inline src/Plane/PlaneFileGlue.cpp:228]: Useless logging call

File: src/Plane/PlaneFileGlue.hpp
Comment (MaxKellermann) [inline src/Plane/PlaneFileGlue.hpp:41]: Why change indentation?

File: src/Plane/PlaneFileGlue.hpp
Comment (MaxKellermann) [inline src/Plane/PlaneFileGlue.hpp:42]: bad indent

File: src/Plane/PlaneFileGlue.hpp
Comment (MaxKellermann) [inline src/Plane/PlaneFileGlue.hpp:43]: Add `[[gnu::pure]]` and `noexcept` to allow more compiler optimizations

File: src/Blackboard/DeviceBlackboard.cpp
Comment (MaxKellermann) [inline src/Blackboard/DeviceBlackboard.cpp:262]: Bad indent

File: src/Blackboard/DeviceBlackboard.cpp
Comment (MaxKellermann) [inline src/Blackboard/DeviceBlackboard.cpp:270]: Bad indent

File: src/Blackboard/DeviceBlackboard.hpp
Comment (MaxKellermann) [inline src/Blackboard/DeviceBlackboard.hpp:169]: Bad indent

File: src/Device/Descriptor.cpp
Comment (MaxKellermann) [inline src/Device/Descriptor.cpp:966]: Stray space before the `)`

File: src/Device/Driver.hpp
Comment (MaxKellermann) [inline src/Device/Driver.hpp:107]: kilogram is not a weight unit, but a mass unit. The unit of weight would be Newton.

File: src/Device/Driver.hpp
Comment (MaxKellermann) [inline src/Device/Driver.hpp:111]: Bad indent

File: src/Device/Driver.hpp
Comment (MaxKellermann) [inline src/Device/Driver.hpp:118]: Baaaaaad indent

File: src/Device/Driver.hpp
Comment (MaxKellermann) [inline src/Device/Driver.hpp:119]: PutEmpty**Mass** but the parameter is called **weight**??

File: src/Device/Driver/XCVario.cpp
Comment (MaxKellermann) [inline src/Device/Driver/XCVario.cpp:34]: Please ... no more useless log messages.

File: src/Device/Driver/XCVario.cpp
Comment (MaxKellermann) [inline src/Device/Driver/XCVario.cpp:188]: Bad coding style

File: src/Device/Driver/XCVario.cpp
Comment (MaxKellermann) [inline src/Device/Driver/XCVario.cpp:328]: Bad coding style

File: src/Device/Driver/XCVario.cpp
Comment (MaxKellermann) [inline src/Device/Driver/XCVario.cpp:343]: Bad coding style

File: src/Device/Driver/XCVario.cpp
Comment (MaxKellermann) [inline src/Device/Driver/XCVario.cpp:353]: Bad coding style

File: src/Device/Driver/XCVario.cpp
Comment (MaxKellermann) [inline src/Device/Driver/XCVario.cpp:361]: Bad coding style

File: src/Device/Driver/XCVario.cpp
Comment (MaxKellermann) [inline src/Device/Driver/XCVario.cpp:358]: Bad coding style

File: src/Device/Driver/XCVario.cpp
Comment (MaxKellermann) [inline src/Device/Driver/XCVario.cpp:359]: Bad indent

File: src/NMEA/ExternalSettings.cpp
Comment (MaxKellermann) [inline src/NMEA/ExternalSettings.cpp:25]: No no, don't add useless log messages everywhere!

File: src/Device/Driver/XCVario.cpp
Comment (MaxKellermann) [inline src/Device/Driver/XCVario.cpp:218]: Bad coding style

File: src/Device/Driver/XCVario.cpp
Comment (MaxKellermann) [inline src/Device/Driver/XCVario.cpp:228]: Bad coding style

File: src/Device/Driver/XCVario.cpp
Comment (MaxKellermann) [inline src/Device/Driver/XCVario.cpp:35]: Why include this header?

File: src/Device/Driver/XCVario.cpp
Comment (MaxKellermann) [inline src/Device/Driver/XCVario.cpp:65]: This looks like you implemented a timeout detection inside this driver, but there is already `OnLinkTimeout()`, which you also use. Don't add unnecessary complexity to the driver.

File: src/Device/Driver/XCVario.cpp
Comment (MaxKellermann) [inline src/Device/Driver/XCVario.cpp:66]: Fields should be documented

File: src/ApplyExternalSettings.cpp
Comment (MaxKellermann) [inline src/ApplyExternalSettings.cpp:89]: Should we really rewrite the whole plane file every time the vario decides to send a new value? What if a different vario repeats the (unmodified) value every second?

**Review by MaxKellermann — CHANGES_REQUESTED** [review]
Lots of bad coding style.
Your commit messages are long enough, but should start with a single line a la "Subsystem: brief summary", followed by an empty line, and then your long commit message. The list of commits looks very confusing if you start it with a half-sentence.

**Comment by MaxKellermann** [comment]
btw. this PR does much more than just water ballast sync. For example, the whole plane mass feature - which triggers a rewrite of the plane file - requires some more thought.
It would have been simpler to merge this if you had just done the water ballast first, but now it's already an entangled big PR with everything.
Next time, don't do everything at once. It makes everything harder for everybody, including you.

**Comment by iltis42** [comment]
> btw. this PR does much more than just water ballast sync. For example, the whole plane mass feature - which triggers a rewrite of the plane file - requires some more thought. It would have been simpler to merge this if you had just done the water ballast first, but now it's already an entangled big PR with everything. Next time, don't do everything at once. It makes everything harder for everybody, including you.

Thanks for the comprehensive comments! 
The coding and documentation issues are all accepted, will work in everything in the next iteration, except the "timeout" story that is actually not a link timeout, actually LinkTimeout() is implemented either, see the override here: 
https://github.com/XCSoar/XCSoar/blob/689791df51bb50ee47ea12be96f81a452cc3b0c5/src/Device/Driver/XCVario.cpp#L218 , but a timeout on application layer to detect explicitly the link is taking $PXCV protocol, as devices might have this NMEA sentences sending (and taking XCVario protocol) stopped for some reason e.g. igc file upload or by explicit by setup (to another protocol), so we need this supervision mechanism on top of the link state to start protocol negotiation, or decide this is legacy protocol version 1. 

And yes, you are right, while altering the plane and crew masses will trigger the issue from #824, alone the exchange of ballast as an absolute value will fix the trouble. So i plan in the next iteration to drop all masses except ballast from this PR. 

Then in a second PR, we may cover crew/plane masses, as the most volatile part of the plane thinking of altering crew's or a FES glider flying with or without battery, so empty mass that should not be mixed up with ballast as this cannot be dumped. Mostly wing area and polar data can be considered as static, but only for the same wingspan, extension's and winglets may or may not get attached to the same plane, so exchange is also desired.

So we are thinking about a third PR to cover exchange of all data from the plane, such as polar data, wing area, name of plane, etc, and in a first iteration this might just go down to the device as it makes more sense, when XCSoar and the device is connected, to the current plane selected in the device. Reason: XCSoar offers a large library and and good overview and the ease of entering lots of data on a touch-screen compared to enter this with just rotary and a push button on the device, users might only want to manage this in XCSoar, and not in the device. If we decide in that XCSoar is the master, altering data in the current plane file of XCSoar might be not be desired.

File: src/Dialogs/Plane/PlaneDetailsDialog.cpp
Comment (iltis42) [reply src/Dialogs/Plane/PlaneDetailsDialog.cpp]: Yup, we need to sent this on Activate either.

File: src/NMEA/ExternalSettings.cpp
Comment (iltis42) [reply src/NMEA/ExternalSettings.cpp]: All logs removed, except one when a wrong protocol element is received.

File: src/Device/Driver/XCVario.cpp
Comment (iltis42) [reply src/Device/Driver/XCVario.cpp]: moved

File: src/Device/Driver/XCVario.cpp
Comment (iltis42) [reply src/Device/Driver/XCVario.cpp]: not clear about the issue here

File: src/Device/Driver/XCVario.cpp
Comment (iltis42) [reply src/Device/Driver/XCVario.cpp]: The Timer callback needs a null environment, in order to declare this.

File: src/Device/Driver/XCVario.cpp
Comment (iltis42) [reply src/Device/Driver/XCVario.cpp]: its actually an application layer XCV protocol supervision and on top of the LinkTimeout()

File: src/Device/Driver/XCVario.cpp
Comment (iltis42) [reply src/Device/Driver/XCVario.cpp]: done

File: src/ApplyExternalSettings.cpp
Comment (iltis42) [reply src/ApplyExternalSettings.cpp]: to be discussed in the next PR, plane masses are removed for now

File: src/Dialogs/Plane/PlaneDetailsDialog.cpp
Comment (iltis42) [reply src/Dialogs/Plane/PlaneDetailsDialog.cpp]: For now plane masses are removed from this PR

File: src/Device/Driver/XCVario.cpp
Comment (iltis42) [reply src/Device/Driver/XCVario.cpp]: fixed

File: src/Device/Driver/XCVario.cpp
Comment (iltis42) [reply src/Device/Driver/XCVario.cpp]: stray space removed after (

File: src/Device/Driver/XCVario.cpp
Comment (iltis42) [reply src/Device/Driver/XCVario.cpp]: stray spaces removed

File: src/Device/Driver/XCVario.cpp
Comment (iltis42) [reply src/Device/Driver/XCVario.cpp]: stray spaces removed

File: src/Device/Driver/XCVario.cpp
Comment (iltis42) [reply src/Device/Driver/XCVario.cpp]: stray spaces removed

File: src/Device/Driver/XCVario.cpp
Comment (iltis42) [reply src/Device/Driver/XCVario.cpp]: stray space removed

File: src/Device/Driver/XCVario.cpp
Comment (iltis42) [reply src/Device/Driver/XCVario.cpp]: stray spaces removed and repetition of constant

File: src/Device/Driver/XCVario.cpp
Comment (iltis42) [reply src/Device/Driver/XCVario.cpp]: just only one log in an error condition

File: src/Device/Driver/XCVario.cpp
Comment (iltis42) [reply src/Device/Driver/XCVario.cpp]: stray spaces removed

File: src/Device/Driver/XCVario.cpp
Comment (iltis42) [reply src/Device/Driver/XCVario.cpp]: stray newline removed

File: src/Device/Driver.hpp
Comment (iltis42) [reply src/Device/Driver.hpp]: will be renamed in plane masses PR

File: src/Device/Driver.hpp
Comment (iltis42) [reply src/Device/Driver.hpp]: yup, to be fixed in next PR for plane masses

File: src/Device/Driver.hpp
Comment (iltis42) [reply src/Device/Driver.hpp]: fixed

File: src/Device/Descriptor.cpp
Comment (iltis42) [reply src/Device/Descriptor.cpp]: will be fixed in plane masses PR

File: src/Plane/PlaneFileGlue.hpp
Comment (iltis42) [reply src/Plane/PlaneFileGlue.hpp]: aligned to the scope of this function with one indent

File: src/Dialogs/Plane/PlaneListDialog.cpp
Comment (iltis42) [reply src/Dialogs/Plane/PlaneListDialog.cpp]: to be done in PR for plane masses

File: src/Dialogs/Plane/PlaneListDialog.cpp
Comment (iltis42) [reply src/Dialogs/Plane/PlaneListDialog.cpp]: to be done in PR for plane masses

File: src/Plane/PlaneFileGlue.cpp
Comment (iltis42) [reply src/Plane/PlaneFileGlue.cpp]: to be done in PR for plane masses

File: src/Plane/PlaneFileGlue.cpp
Comment (iltis42) [reply src/Plane/PlaneFileGlue.cpp]: will be removed in PR for plane masses

File: src/Plane/PlaneFileGlue.cpp
Comment (iltis42) [reply src/Plane/PlaneFileGlue.cpp]: to be added in PR for plane masses

File: src/Plane/PlaneFileGlue.cpp
Comment (iltis42) [reply src/Plane/PlaneFileGlue.cpp]: Sure, will be changed in PR for plane masses

File: src/Plane/PlaneFileGlue.cpp
Comment (iltis42) [reply src/Plane/PlaneFileGlue.cpp]: Used during test, will be removed in PR for plane masses

File: src/Plane/PlaneFileGlue.hpp
Comment (iltis42) [reply src/Plane/PlaneFileGlue.hpp]: Sure, to be aligned in PR for plane masses

File: src/Blackboard/DeviceBlackboard.cpp
Comment (iltis42) [reply src/Blackboard/DeviceBlackboard.cpp]: to be aligned in PR for plane masses

File: src/Blackboard/DeviceBlackboard.hpp
Comment (iltis42) [reply src/Blackboard/DeviceBlackboard.hpp]: to be aligned in PR for plane masses

File: src/Device/Driver.hpp
Comment (iltis42) [reply src/Device/Driver.hpp]: weight needs to be replaced by mass in PR for plane masses

File: src/Blackboard/DeviceBlackboard.cpp
Comment (iltis42) [reply src/Blackboard/DeviceBlackboard.cpp]: to be aligned in PR for plane masses

File: src/Device/Driver/XCVario.cpp
Comment (MaxKellermann) [reply src/Device/Driver/XCVario.cpp]: This is bad, because it means that `OnSysTicker()` can block, but it must not. By abusing a `NullOperationEnvironment`, you escape the protection against this mistake I built into the API.

File: src/Device/Driver/XCVario.cpp
Comment (MaxKellermann) [reply src/Device/Driver/XCVario.cpp]: I don't understand. This needs further explanation.

Comment (iltis42) [reply src/Device/Driver/XCVario.cpp]: Once the link comes up, XCVario protocol needs further negotiation on application layer what protocol to run. To be backward compatible with older device SW, there is protocol number negotiation between XCS and the XCV device, and until this hasn't been finished, we can't clearly go into full operation, so there is a second decision on top of the link status below on application layer to decide about this. Older devices might not answer the protocol version query, so after the link is up and if within a timeout of 5 seconds there is no answer, XCS will run in legacy mode with fractional ballasts. And until then, as long as the protocol version is unknown, any sending of NMEA except protocol version negotiation, must not take place.

**Pull request closed without merging**
